martian/src/mro.rs

```diff
@@ -284,24 +320,26 @@ impl<T: AsMartianBlanketType> AsMartianBlanketType for Option<T> {
 
 impl<T: AsMartianBlanketType> AsMartianBlanketType for Vec<T> {
     fn as_martian_blanket_type() -> MartianBlanketType {
-        match T::as_martian_blanket_type() {
-            MartianBlanketType::Primary(primary) => MartianBlanketType::Array(primary),
-            MartianBlanketType::Array(_) => {
-                unimplemented!("Array of arrays are not supported in martian")
-            }
-        }
+        MartianBlanketType::Array(Box::new(T::as_martian_blanket_type()))
     }
 }
 
 impl<K: AsMartianPrimaryType, H> AsMartianBlanketType for HashSet<K, H> {
     fn as_martian_blanket_type() -> MartianBlanketType {
-        MartianBlanketType::Array(K::as_martian_primary_type())
+        MartianBlanketType::Array(Box::new(K::as_martian_blanket_type()))
     }
 }
 
-impl<K, V, H> AsMartianPrimaryType for HashMap<K, V, H> {
-    fn as_martian_primary_type() -> MartianPrimaryType {
-        MartianPrimaryType::Map
+// ideally we'd allow for any HashMap to be turned into a typed Map when possible, or an untyped Map by default
+// but a typed map can only be made for Rust HashMap with a key implementing Display + Eq + Hash
+// and a value implementing MartianPrimaryType
+// it is not possible to have multiple implementations ranked in priority without specialization, which is an unstable feature
+// and it is impossible to check what traits are implemented for a HashMap's K,V at runtime.
+// instead, the current solution is that any HashMap not meeting these trait bounds must manually specify the type
+// using #[mro_type = "map"]
+impl<K, V: AsMartianBlanketType, H> AsMartianBlanketType for HashMap<K, V, H> {
```

Comment:
I would keep `HashMap` as an untyped generic map with no restrictions and define a struct `TypedMap<K, V, H>` which is a simple wrapper around `HashMap<K, V, H>` where `V: AsMartianBlanketType`. This preserves backward compatibility but also makes typed map an opt in. What do you think about that?

Reply:
Hmmm. I tried this strategy earlier but found it not very ergonomic to implement. The least annoying way would involve having TypedMap implement `Deref` and `DerefMut` so that it can effectively "inherit" the functionality of the contained map without having to unwrap it or implement a suite of accessors. But I think I ran into some trouble doing this because of the HashMap's generic type parameters. Further, it would become necessary to pass around the Hasher we need unless we decide to abstract them away as with `TxHashMap`. Would you recommend we also create a `TypedTxHashMap` newtype in `metric` and make `Martian` a new dependency of `metric`?

Reply:
Personally, I was of the same mind as you, but the fact that I only had to manually untype 2 or 3 Maps in the entire Cellranger pipeline implies this isn't the worst breaking change we could have. And I think @adam-azarchs was pushing earlier for typed maps by default.

Reply:
Ah, if it's just 2 or 3 untyped maps in all of cellranger, that's not bad.

Reply:
Best practices should always be on by default.  Untyped map should be opt-in and strongly discoraged.

Reply:
Really the only case where an untyped map should ever be produced is where martian doesn't allow one to express the type constraint, e.g. where the values are an `enum` type.  Though in most such cases what you should probably actually have is a `struct`.